backend/src/main/scala/com/softwaremill/adopttapir/starter/FilesGenerator.scala

```diff
@@ -0,0 +1,53 @@
+package com.softwaremill.adopttapir.starter
+
+import com.softwaremill.adopttapir.template.{GeneratedFile, SbtProjectTemplate, ScalaCliProjectTemplate}
+import com.softwaremill.adopttapir.template.SbtProjectTemplate.legalizeGroupId
+
+object FilesGenerator {
+  implicit class StarterDetailsWithFilesGenerator(starterDetails: StarterDetails) {
+    def generateFiles: List[GeneratedFile] = {
+      starterDetails.builder match {
+        case Builder.Sbt      => SbtFilesGenerator.generate(starterDetails)
+        case Builder.ScalaCli => ScalaCliFilesGenerator.generate(starterDetails)
+      }
+    }
+  }
+}
+
+trait FilesGenerator {
+  def generate(starterDetails: StarterDetails): List[GeneratedFile]
+}
+
+private case object SbtFilesGenerator extends FilesGenerator {
+  private lazy val sbtTemplate: SbtProjectTemplate = new SbtProjectTemplate()
+
+  override def generate(starterDetails: StarterDetails): List[GeneratedFile] = {
+    List(
+      sbtTemplate.getBuildSbt(starterDetails),
+      sbtTemplate.getBuildProperties,
+      sbtTemplate.getMain(legalizeGroupId(starterDetails)),
+      sbtTemplate.getEndpoints(legalizeGroupId(starterDetails)),
+      sbtTemplate.getEndpointsSpec(legalizeGroupId(starterDetails)),
+      sbtTemplate.pluginsSbt,
+      sbtTemplate.scalafmtConf(starterDetails.scalaVersion),
+      sbtTemplate.sbtx,
+      sbtTemplate.README
+    )
+  }
+}
+
+private case object ScalaCliFilesGenerator extends FilesGenerator {
+  private lazy val scalaCliTemplate: ScalaCliProjectTemplate = new ScalaCliProjectTemplate()
+
+  override def generate(starterDetails: StarterDetails): List[GeneratedFile] = {
+    List(
+      scalaCliTemplate.getBuildScalaCli(starterDetails),
+      scalaCliTemplate.getTestScalaCli(starterDetails),
+      scalaCliTemplate.getMain(legalizeGroupId(starterDetails)),
```

Comment:
Method SbtProjectTemplate#legalizeGroupId should be moved to ProjectTemplate as it is used by both generators. 

Maybe make it private and run under body of .getMain / .getEndpoints / .getEndpointsSpec .

Reply:
good point, done!